src/contracts/core/Slasher.sol

```diff
@@ -5,233 +5,475 @@ import "./SlasherStorage.sol";
+
+    /// EVENTS
+    
+    event QueuedAllocation(
+        address operator,
+        IStrategy strategy,
+        OperatorSet operatorSet,
+        uint32 effectTimestamp,
+        uint64 slashableMagnitude
+    );
+    
+    event QueuedDeallocation(
+        address operator,
+        IStrategy strategy,
+        OperatorSet operatorSet,
+        uint32 effectTimestamp,
+        uint64 slashableMagnitude
+    );
+
+    event TotalAndNonSlashableMagnitudeUpdated(
+        address operator,
+        IStrategy strategy,
+        uint32 effectTimestamp,
+        uint64 nonSlashableMagnitude,
+        uint64 totalSlashableMagnitude
+    );
+
+    /// operator => strategy => avs => operatorSetId => MagnitudeUpdate[]
+    mapping(address => mapping(IStrategy => mapping(address => mapping(uint32 => MagnitudeUpdate[])))) private _magnitudeUpdates;
+
+    /// operator => strategy => TotalAndNonslashableUpdate[]
+    mapping(address => mapping(IStrategy => TotalAndNonslashableUpdate[])) private _totalMagnitudeUpdates;
+    /// operator => allocator
+    /// could perhaps use 
+    mapping(address => address) public allocatorFor;
+
     constructor(
         IStrategyManager _strategyManager,
-        IDelegationManager _delegationManager,
-        IOperatorSetManager _operatorSetManager
-    ) SlasherStorage(_strategyManager, _delegationManager, _operatorSetManager) {}
+        IDelegationManager _delegationManager
+    ) SlasherStorage(_strategyManager, _delegationManager) {}
 
     /**
-     * @notice Called by an AVS to increase its own slashing request for a given
-     * operator set and operator in the current epoch
+     * @notice Queues slashing magnitude deallocation for an 
+     * (operator, strategy, operatorSet) tuple.
+     * Queues magnitude deallocation 21 days from now which will be automatically applied
+     * at that respective time.
+     * 
+     * @param operator the operator whom the slashing parameters are being changed
+     * @param deallocationParams deallocation params with differences to subtract from current magnitudes
+     * @param allocatorSignature if non-empty is the signature of the allocator on 
+     * the modification. if empty, the msg.sender must be the allocator for the 
+     * operator
      *
-     * @param operator the operator that the calling AVS is to increase the bips they want to slash
-     * @param operatorSetID the id of the operator set the AVS is increasing their slashing for
-     * @param strategies the list of strategies slashing requested is being modified for
-     * @param bipsToIncrease the basis points slashing to modify for given strategies
+     * @dev pushes a MagnitudeUpdate and TotalAndNonslashableUpdate to take effect in 21 days
+     * @dev reverts if magnitudeDiff > allocated magnitude
+     * since you cannot deallocate more than is already allocated
+     * @dev emits event queueDeallocation
      */
-    function increaseRequestedBipsToSlash(
+    function queueDeallocation(
```

Comment:
lowkey should we just do everything in ints so we dont duplicate this code twice

Reply:
Tried to think of best way to reuse some kind of internal function.
Main thing with this is that addedMagnitude check is comparing with nonslashableMagnitude
and decrementedMagnitude check is comparing with latest allocated magnitude.
Handling these two different cases between allocation/deallocation seemed difficult so I left the params as uint64